src/machine/machine_avr.go

```diff
@@ -271,3 +272,122 @@ func (i2c I2C) ReadByte() byte {
 
 	return byte(*avr.TWDR)
 }
+
+// UART
+
+type UARTConfig struct {
+	Baudrate uint32
+}
+
+type UART struct {
+}
+
+var (
+	// UART0 is the hardware serial port on the AVR.
+	UART0 = &UART{}
+)
+
+// Configure the UART on the AVR.
+func (uart UART) Configure(config UARTConfig) {
+	// default to 115200
+	if config.Baudrate == 0 {
+		config.Baudrate = 115200
+	}
+
+	// Set baud rate based on prescale formula:
+	// BAUD_PRESCALE = (((F_CPU / (BAUDRATE * 16UL))) - 1)
+	ps := (CPU_FREQUENCY / (config.Baudrate * 16)) - 1
+	*avr.UBRR0H = avr.RegValue(ps >> 8)
+	*avr.UBRR0L = avr.RegValue(ps)
+
+	// enable RX interrupt
+	*avr.UCSR0B |= avr.UCSR0B_RXCIE0
+}
+
+// Close the UART on the AVR.
+func (uart UART) Close() error {
+	return nil
+}
+
+// Read from the RX buffer.
+func (uart UART) Read(data []byte) (n int, err error) {
+	if len(data) > RXBufferSize {
+		return 0, errors.New("Read buffer cannot be larger than RXBuffer")
+	}
+
+	// check if RX buffer is empty
+	size := BufferUsed()
+	if size == 0 {
+		return 0, nil
+	}
+
+	// only read number of bytes used from buffer
+	for i := 0; uint8(i) < size; i++ {
+		data[i] = byte(bufferGet())
+	}
+
+	return len(data), nil
+}
+
+// Write data to the UART.
+func (uart UART) Write(data []byte) (n int, err error) {
+	for _, v := range data {
+		uart.WriteByte(v)
+	}
+	return len(data), nil
+}
+
+// ReadByte reads a single byte from the RX buffer.
+func (uart UART) ReadByte() (byte, error) {
+	return byte(bufferGet()), nil
+}
+
+// WriteByte writes a byte of data to the UART.
+func (uart UART) WriteByte(c byte) error {
+	// Wait until UART buffer is not busy.
+	for (*avr.UCSR0A & avr.UCSR0A_UDRE0) == 0 {
+	}
+	*avr.UDR0 = avr.RegValue(c) // send char
+	return nil
+}
+
+type __volatile byte
+
+const RXBufferSize = 64
+
+// Minimal ring buffer implementation inspired by post at
+// https://www.embeddedrelated.com/showthread/comp.arch.embedded/77084-1.php
+var rxbuffer [RXBufferSize]__volatile
+var head __volatile
+var tail __volatile
+
+func BufferUsed() uint8 {
+	return uint8(bufferUsed())
+}
+
+func bufferUsed() __volatile { return head - tail }
+func bufferPut(val __volatile) {
+	if bufferUsed() != RXBufferSize {
+		head++
+		rxbuffer[head%RXBufferSize] = val
+	}
+}
+func bufferGet() __volatile {
+	if bufferUsed() != 0 {
+		tail++
+		return rxbuffer[tail%RXBufferSize]
+	}
+	return 0
```

Comment:
I'm not entirely convinced ReadByte should not block. In any case, this function should somehow either block itself or indicate to the caller it would block (for example, by returning an extra 'ok' bool) so that a user of ReadByte can receive binary data.

Reply:
From GoDocs, "ReadByte reads and returns a single byte. If no byte is available, returns an error."

https://golang.org/pkg/bufio/#Reader.ReadBytes

Reply:
That was the wrong link, but the correct description of the function. So my current implementation is what the interface expects: https://github.com/aykevl/tinygo/pull/25/files#diff-f30c9b1e4e77d9fd02ae4156f2cfc809R338

Reply:
Also the implementation works fine for binary data.

Reply:
Oh, now I see. You are right. Read() and ReadByte() both check Buffered() first, so they should be able to handle binary data (including null bytes).